### PR TITLE
Task to add eBOSS-like redshift errors to catalog

### PR DESCRIPTION
**Pull request by sjforeman**
This refactors Seth's eBOSS redshift error code into a separate task, and implements a few other minor tweaks.

### CODE REVIEW

File: draco/synthesis/mockcatalog.py [hunk 1]
```diff
@@ -795,19 +809,213 @@ def process(self, cat):
...
+        err_func = _velocity_error_function_lookup[tracer]
+
+        dv = err_func(z.size, self.rng)
+
+        dz = (1.0 + z) * dv / (C * 1e-3)
+
+        return dz
+
+    @staticmethod
+    def _qso_velocity_error(nsample, rng):
```
Comment (jrs65) [inline draco/synthesis/mockcatalog.py:919]: This all looks great. My only comment would be to remove the `_` prefixes from these so they're no longer "private" and it's clearer that they could be used elsewhere.

File: draco/synthesis/mockcatalog.py [hunk 2]
```diff
@@ -753,16 +759,23 @@ def process(self):
...
+    catalog.
 
-    TODO: allow for user-specified PDF.
+    Note that the errors are added to the catalog in place, such that
+    if the original catalog is subsequently used in a pipeline, it will
+    have the errors included.
 
     Attributes
     ----------
+    use_catalog_z_errors : bool
```
Comment (jrs65) [inline draco/synthesis/mockcatalog.py:775]: Yeah, this seems like a good option to have.

File: draco/synthesis/mockcatalog.py [hunk 3]
```diff
@@ -795,19 +809,213 @@ def process(self, cat):
...
+        cat_z_err = cat["redshift"]["z_error"][:]
+
+        # Generate redshift errors for the chosen tracer
+        z_err = self._generate_z_errors(cat_z, self.tracer)
 
         # Add errors to catalog redshifts
         cat_z += z_err
         # Add errors in quadrature with existing stored errors
-        cat_z_err = (z_err ** 2 + cat_z_err ** 2) ** 0.5
+        cat_z_err[:] = (z_err ** 2 + cat_z_err ** 2) ** 0.5
```
Comment (jrs65) [inline draco/synthesis/mockcatalog.py:875]: It's not obvious this makes any sense. `z_err` is a realisation of a random variable, but you want to add in an estimate of the distribution variance instead. That probably requires another method to calculate that.

**Review by jrs65 — CHANGES_REQUESTED** [review]
A few small changes, otherwise very nice!

File: draco/synthesis/mockcatalog.py (hunk 1 above)
Comment (sjforeman) [reply draco/synthesis/mockcatalog.py]: Changed!

**Review by ssiegelx — APPROVED** [review]
This looks great, thanks for pulling it all together.  I just have one comment.

File: draco/synthesis/mockcatalog.py (hunk 3 above)
Comment (sjforeman) [reply draco/synthesis/mockcatalog.py:875]: Agreed - this somehow snuck through a previous review, and I left it alone as I wasn't sure what would be best here. Do we want to save the standard deviation of the underlying distribution even if the distribution isn't Gaussian? What is the intended use of the `z_error` field - simply a reflection of a (Gaussian) errorbar on `z`?

File: draco/synthesis/mockcatalog.py [hunk 4]
```diff
@@ -795,19 +809,213 @@ def process(self, cat):
...
+    if the original catalog is subsequently used in a pipeline, it will
+    have the errors included.
+
+    Attributes
+    ----------
+    tracer : {"ELG"|"LRG"|"QSO"}
+        Generate redshift errors corresponding to this eBOSS sample.
+    """
+
+    tracer = config.enum(["QSO", "ELG", "LRG"])
```
[the remark this answers is not in the document]
Comment (jrs65) [reply draco/synthesis/mockcatalog.py:849]: That sounds very useful, although fwiw it wouldn't actually help the way I run things (I use `caput-pipeline template-run` to have the tracer be a variable within a template that gets filled in and run for each tracer type).

File: draco/synthesis/mockcatalog.py (hunk 3 above)
Comment (jrs65) [reply draco/synthesis/mockcatalog.py:875]: The point of the `z_error` field was simply that it was a column available in the SDSS database and it sounded useful to have in our analysis, but I don't think we've ever actually used it, and so we've no real restriction of what it should be.

I think in this case, probably the best thing would be to estimate an effective sigma for each type (even for the non-Gaussian distributions), but I don't think we're really going to need it immediately so I'd be happy just not modifying it in this task and making an issue to do something smarter at a future date.

File: draco/synthesis/mockcatalog.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jrs65) [reply draco/synthesis/mockcatalog.py:849]: Just to be sure, maybe I'll modify the Mockcatalog code to add a tracer attribute anyway before I generate all the catalogs.

File: draco/synthesis/mockcatalog.py (hunk 3 above)
Comment (sjforeman) [reply draco/synthesis/mockcatalog.py]: Sure, sounds good to me

File: draco/synthesis/mockcatalog.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (sjforeman) [reply draco/synthesis/mockcatalog.py]: Thanks Seth - I inadvertently dropped this functionality from your code, so thanks for reminding me to put it back in!

File: draco/synthesis/mockcatalog.py (hunk 3 above)
Comment (sjforeman) [reply draco/synthesis/mockcatalog.py]: See https://github.com/radiocosmology/draco/issues/151

File: draco/synthesis/mockcatalog.py [hunk 5]
```diff
@@ -795,20 +809,236 @@ def process(self, cat):
...
+
+    Note also that the errors are added to the catalog in place, such that
+    if the original catalog is subsequently used in a pipeline, it will
+    have the errors included.
+
+    Attributes
+    ----------
+    tracer : {"ELG"|"LRG"|"QSO"}
+        Generate redshift errors corresponding to this eBOSS sample.
+        If not specified, task will attempt to detect the tracre type from
```
Comment (jrs65) [inline draco/synthesis/mockcatalog.py:845]:
```suggestion
        If not specified, task will attempt to detect the tracer type from
```

File: draco/synthesis/mockcatalog.py [hunk 6]
```diff
@@ -795,20 +809,236 @@ def process(self, cat):
...
+                for key in _velocity_error_function_lookup.keys():
+                    if key in cat.attrs["tag"]:
+                        tracer = key
+                        break
+
+                if tracer is None:
+                    raise ValueError(
+                        "Must specify eBOSS tracer in config property, "
+                        "catalog 'tracer' attribute, or catalog 'tag' attribute."
+                    )
```
Comment (jrs65) [inline draco/synthesis/mockcatalog.py:882]: Maybe make this detection case insensitive? We've tended to use a bit of a mixed version of the tracer abbreviations.

Also check that the tracer attribute is actually understood.
```suggestion
        if tracer is None:
            if "tracer" in cat.attrs:
                tracer = cat.attrs["tracer"].upper()
                
                if tracer not in _velocity_error_function_lookup:
                    raise ValueError(
                        f"Tracer explicitly set to '{tracer}' in catalog, but value not supported."
                    )
            else:
                for key in _velocity_error_function_lookup.keys():
                    if key in cat.attrs["tag"].upper():
                        tracer = key
                        break

                if tracer is None:
                    raise ValueError(
                        "Must specify eBOSS tracer in config property, "
                        "catalog 'tracer' attribute, or catalog 'tag' attribute."
                    )
```

**Review by jrs65 — APPROVED** [review]
A few minor suggestions, but you can probably ignore them if you want.

**Comment by jrs65** [comment]
Looks good to me!

**Pull request merged**